walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,93 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+
 
 public final class FindMeetingQuery {
+
+  // create a set of events to avoid duplicates
+  private Collection<Event> events = new HashSet<>();
+
+  // the meeting request
+  private MeetingRequest request; 
+
+  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
+  private static final Collection<String> NO_ATTENDEES = Collections.emptySet();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+ 
+    this.events = events;
+    this.request = request; 
+
+    // Check if duration is longer than a day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList(); 
+    }
+    // Check for no events or no attendees
+    if (events == NO_EVENTS || request.getAttendees() == NO_ATTENDEES) {
+        return Arrays.asList((TimeRange.WHOLE_DAY)); 
+    }
+    // Arraylist holds event times accordingly     
+    List<TimeRange> mandatory = new ArrayList();
+
+    for (Event event: events) {
+
+        // Check if attendees from event are requested for meeting
+        for (String attendee : event.getAttendees()) {            
+            // Check that attendee is a mandatory attendee
+            if ((request.getAttendees().contains(attendee)) ) {
+                mandatory.add(event.getWhen());
+            }           
+        }
+    }
+
+    // Create empty Arraylist of options
+    List<TimeRange> options = new ArrayList();
+
+    // Check if there are valid events during the day
+    if (mandatory.isEmpty()) {
+        return Arrays.asList((TimeRange.WHOLE_DAY)); 
+    }
+
+    // Sort list of time ranges in ascending order
+    Collections.sort(mandatory, TimeRange.ORDER_BY_START);
+
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.END_OF_DAY; 
+    int lastEventEnd = 0;
+
+    for (int i = 0; i <= (mandatory.size() -1); i++)  {
+        // Get event start and end time
+        int eventStart = mandatory.get(i).start();
+        int eventEnd = mandatory.get(i).end(); 
+    
+        // Check for overlap
+        if (start < eventStart) {
+            // Check if there is enough room for a meeting
+            if ((eventStart - start) >= request.getDuration()) {
+                options.add(TimeRange.fromStartEnd(start,eventStart,false)); 
+            }
+        }  
+        start = Math.max(start, eventEnd);
+        lastEventEnd = start;
+    } 
+
+    // Check if last event ends when day ends
+    if (lastEventEnd != end+1) {
+        // Check if there is enough room for a meeting
+        if ((end+1 - lastEventEnd) >= request.getDuration()) {
+            options.add(TimeRange.fromStartEnd(lastEventEnd,end,true)); 
```

Comment:
If a TimeRange is non-inclusive range, the end value here should be end+1.

But this looks overall like there's an off-by-one error in the code somewhere.

Reply:
Which specific test is failing if you don't have this special case?

Reply:
justEnoughRoom (would return an extra Range [1440,1440) ) and notEnoughRoom (returns Range [1440,1440) when expected is [] ) These are both because of the off-by-one difference of the TimeRange.END_OF_DAY vs the event.end()

Reply:
How would the range [1440,1440) pass the check where it's >= the request.getDuration() ?

It seems like it would be a range of length 0 and should fail.

What about changing line 97 to be (lastEventEnd < END_OF_DAY)?